### PR TITLE
 DL: Add new param metrics_compute_frequency to madlib_keras_fit()

### PR DESCRIPTION
**Pull request by njayaram2**
JIRA: MADLIB-1335
This commit adds a new optional parameter `metrics_compute_frequency` to
`madlib_keras_fit()` to control how often Keras.evaluate() is run on
training and validation data for computing loss/metrics. We only support
one metric at the moment.

This is a WIP PR, more work will be done after changes for MADLIB-1332 is pushed to master.

### CODE REVIEW

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/895/

File: src/ports/postgres/modules/deep_learning/madlib_keras.py_in [hunk 1]
```diff
@@ -173,134 +179,131 @@ def fit(schema_madlib, source_table, model,model_arch_table,
...
+            $1 AS compile_params,
+            $2 AS fit_params,
+            {num_iterations} AS num_iterations,
+            {validation_table}::TEXT AS validation_table,
+            {metrics_compute_frequency} AS metrics_compute_frequency,
+            $3 AS name,
+            $4 AS description,
+            '{model_type}'::TEXT AS model_type,
+            {model_size} AS model_size,
+            '{start_training_time}'::TIMESTAMP AS start_training_time,
```
Comment (kaknikhil) [inline src/ports/postgres/modules/deep_learning/madlib_keras.py_in]: do we really need to quote the timestamp columns ?

File: src/ports/postgres/modules/deep_learning/madlib_keras.py_in [hunk 2]
```diff
@@ -313,6 +316,61 @@ def fit(schema_madlib, source_table, model,model_arch_table,
...
+                                                   model_arch, model_state,
+                                                   gpus_per_host,
+                                                   segments_per_host,
+                                                   seg_ids_val,
+                                                   rows_per_seg_val,
+                                                   gp_segment_id_col)
+    end_val = time.time()
+    plpy.info("Time for evaluation in iteration {0}: {1} sec.". format(
+        curr_iter + 1, end_val - start_val))
+    if len(evaluate_result) < 2:
```
Comment (kaknikhil) [inline src/ports/postgres/modules/deep_learning/madlib_keras.py_in]: metric is an optional param which means that len(evaluate_result) can be >=1.

File: src/ports/postgres/modules/deep_learning/madlib_keras.py_in [hunk 3]
```diff
@@ -313,6 +316,61 @@ def fit(schema_madlib, source_table, model,model_arch_table,
...
+    loss = evaluate_result[0]
+    metric = evaluate_result[1]
+    plpy.info("{0} set metric after iteration {1}: {2}.".
+              format(dataset_name, curr_iter + 1, metric))
+    plpy.info("{0} set loss after iteration {1}: {2}.".
+              format(dataset_name, curr_iter + 1, loss))
+    metrics_list.append(metric)
+    loss_list.append(loss)
+
+def should_compute_metrics_this_iter(
```
Comment (kaknikhil) [inline src/ports/postgres/modules/deep_learning/madlib_keras.py_in]: I think it is more readable if the current iteration is one indexed

File: src/ports/postgres/modules/deep_learning/test/madlib_keras.sql_in [hunk 4]
```diff
@@ -198,37 +228,42 @@ SELECT madlib_keras_fit(
     2,
     NULL,
     NULL,
+    NULL,
     'model name', 'model desc');
+\x on;
+select * from keras_out_summary;
```
Comment (kaknikhil) [inline src/ports/postgres/modules/deep_learning/test/madlib_keras.sql_in]: lines 233-234 look like leftover debug code.

File: src/ports/postgres/modules/deep_learning/madlib_keras.py_in [hunk 5]
```diff
@@ -173,134 +179,131 @@ def fit(schema_madlib, source_table, model,model_arch_table,
...
     norm_const, norm_const_type = get_col_value_and_type(
         fit_validator.source_summary_table, NORMALIZING_CONST_COLNAME)
     dep_vartype = plpy.execute("SELECT {0} AS dep FROM {1}".format(
         DEPENDENT_VARTYPE_COLNAME, fit_validator.source_summary_table))[0]['dep']
-    dependent_varname_in_source_table = quote_ident(plpy.execute("SELECT {0} FROM {1}".format(
-        'dependent_varname', fit_validator.source_summary_table))[0]['dependent_varname'])
-    independent_varname_in_source_table = quote_ident(plpy.execute("SELECT {0} FROM {1}".format(
-        'independent_varname', fit_validator.source_summary_table))[0]['independent_varname'])
+    # Quote_ident TEXT values to be inserted into the summary table
+    dependent_varname_in_source_table = plpy.execute("SELECT {0} FROM {1}".format(
```
Comment (kaknikhil) [inline src/ports/postgres/modules/deep_learning/madlib_keras.py_in]: Consider creating a function for creating the two output tables. The only downside is that we might have to pass in a lot of arguments to this function. I will leave the decision up to you.

File: src/ports/postgres/modules/deep_learning/madlib_keras.py_in [hunk 6]
```diff
@@ -66,7 +66,12 @@ def fit(schema_madlib, source_table, model,model_arch_table,
 
     fit_validator = FitInputValidator(
         source_table, validation_table, model, model_arch_table,
-        dependent_varname, independent_varname, num_iterations)
+        dependent_varname, independent_varname, num_iterations,
+        metrics_compute_frequency)
+    # If metrics_compute_frequency is NULL, assign num_iterations as
```
Comment (kaknikhil) [inline src/ports/postgres/modules/deep_learning/madlib_keras.py_in]: I think the code is self explanatory here, don't really need a comment

File: src/ports/postgres/modules/deep_learning/madlib_keras.py_in (hunk 2 above)
Comment (njayaram2) [reply src/ports/postgres/modules/deep_learning/madlib_keras.py_in]: Yes, this will be addressed as part of https://issues.apache.org/jira/browse/MADLIB-1338 I guess.

File: src/ports/postgres/modules/deep_learning/madlib_keras.py_in (hunk 3 above)
Comment (njayaram2) [reply src/ports/postgres/modules/deep_learning/madlib_keras.py_in]: Will change it in the fit function too, it will be inconsistent otherwise. For instance, all the prints we have do an `i+1` while reporting numbers, and `compute_loss_and_metrics` also assumes zero indexed `i` value. Will instead run the for loop in `range(1, num_iterations+1)` and make it consistent everywhere.

File: src/ports/postgres/modules/deep_learning/madlib_keras.py_in (hunk 1 above)
Comment (njayaram2) [reply src/ports/postgres/modules/deep_learning/madlib_keras.py_in]: Yes, the query fails otherwise.

File: src/ports/postgres/modules/deep_learning/madlib_keras.py_in (hunk 5 above)
Comment (njayaram2) [reply src/ports/postgres/modules/deep_learning/madlib_keras.py_in]: I think it's better if we do this and a bunch of other summary table reads in one shot and consume it. Refactored the code to do that.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/900/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/901/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/902/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/905/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/906/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/909/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/911/
